include/deal.II/base/ndarray.h

```diff
@@ -0,0 +1,102 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_ndarray_h
+#define dealii_ndarray_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+#include <array>
+
+DEAL_II_NAMESPACE_OPEN
+
+#ifndef DOXYGEN
+namespace
+{
+  // clang-format off
+  //
+  // A variadic template helper class to recursively "unroll" the size
+  // information of the ndarray. This is best explained on an example:
+  //
+  //    HelperArray<double, 1, 2, 3, 4>::type
+  // == std::array<HelperArray<double, 2, 3, 4>::type, 1>
+  // == std::array<std::array<HelperArray<double, 3, 4>::type, 2>, 1>
+  // == std::array<std::array<std::array<HelperArray<double, 4>::type, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<HelperArray<double>::type, 4>, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<double, 4>, 3>, 2>, 1>
+  //
+  // clang-format on
+  template <typename T, std::size_t... Ns>
+  struct HelperArray
+  {};
+
+  // Recursively define the type typedef of HelperArray<T, N, ...Ns> by
+  // wrapping a std::array around HelperArray<T, Ns...>::type
+  template <typename T, std::size_t N, std::size_t... Ns>
+  struct HelperArray<T, N, Ns...>
+  {
+    using type = std::array<typename HelperArray<T, Ns...>::type, N>;
+  };
+
+  // End recursion once no std::size_t template parameters are left and
+  // simply set the type typedef to type T
+  template <typename T>
+  struct HelperArray<T>
+  {
+    using type = T;
+  };
+} // namespace
+#endif // DOXYGEN
+
+/**
+ * A (variadic template) typedef for conveniently defining multidimensional
+ * <a href="https://en.cppreference.com/w/cpp/container/array">std::array</a>s
+ *
+ * The problem we try to address with the typedef is the following. Suppose
+ * you want to create a multdimensional array of doubles of , for example,
+ * rank 3, with sizes 2, 3, 4 for the first, middle, and last index. Then
+ * using C-style arrays you could simply write
+ * @code
+ *   double my_array[2][3][4] = { ... };
+ * @endcode
+ * Nowadays, there are a number of good reasons why using C-style arrays is
+ * usually discouraged (ranging from incompatibilities with STL functions
+ * requiring awkward wrappers, surprises when comparing for equality, etc.)
+ * If you want to do the same, however, using the more modern (and
+ * encouraged) `std::array` then you would have to declare
+ * @code
+ *   std::array<std::array<std::array<double, 4>, 3>, 2> = { ... };
+ * @endcode
+ * The repetitions of `std::array` look awkward and, worse, the index
+ * ranges have reversed: the leftmost index has range [0,2), the middle
+ * index has rnage [0,3) and the rightmost index has range [0,4).
+ * We address this issue by providing a typedef ndarray that allows to you
```

Comment:
```suggestion
 * We address this issue by providing a class `ndarray` that allows to you
```
(It's not exactly a class, but it points to a class, so let's not worry too much...

Reply:
I changed this. "class" is linguistically nicer but what is really going on is that we provide a type alias :-)